### PR TITLE
[cmd/opampsupervisor] Fix early exits not shutting down client/server in bootstrapping

### LINKED ISSUE #31943: [cmd/opampsupervisor] Errors during bootstrapping will fail to shut down collector
### Component(s)

cmd/opampsupervisor

## Description

If collector bootstrapping fails, the Stop method will not be called for the collector command. This causes the collector to outlive the supervisor, which ends up blocking any attempts for the supervisor to start up again (collector occupies port 8888)

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/011256c9ba76bac0d55bed1e1d414572dc8073c5/cmd/opampsupervisor/supervisor/supervisor.go#L308

Solution would be to use `defer` statements to ensure the command is always stopped before returning from the `getBootstrapInfo` method.

### Collector version

f8627a7041d7c1bfda9b641d35265a10e8e1ffc0

## Environment
OS: macos
Compiler(if manually compiled): go 1.22

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
**Description:**
* Fix early exits not shutting down client/server in bootstrapping

**Link to tracking Issue:** Closes #31943

**Testing:**
I tested manually adding errors into the code. There isn't a good, consistent way to add e.g. the timeout error as-is.

### CODE REVIEW

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 1]
```diff
@@ -278,6 +278,12 @@ func (s *Supervisor) getBootstrapInfo() (err error) {
 		return err
 	}
 
+	defer func() {
+		if stopErr := srv.Stop(context.Background()); stopErr != nil {
+			err = errors.Join(err, fmt.Errorf("stop opamp server: %w", stopErr))
```
Comment (crobert-1) [inline cmd/opampsupervisor/supervisor/supervisor.go:283]:
```suggestion
			err = errors.Join(err, fmt.Errorf("error when stopping the opamp server: %w", stopErr))
```
Nit

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 2]
```diff
@@ -291,6 +297,12 @@ func (s *Supervisor) getBootstrapInfo() (err error) {
 		return err
 	}
 
+	defer func() {
+		if stopErr := cmd.Stop(context.Background()); stopErr != nil {
+			err = errors.Join(err, fmt.Errorf("stop collector command: %w", stopErr))
```
Comment (crobert-1) [inline cmd/opampsupervisor/supervisor/supervisor.go:302]:
```suggestion
			err = errors.Join(err, fmt.Errorf("error when stopping the collector: %w", stopErr))
```
Nit

**Review by crobert-1 — APPROVED** [review]
Just a couple wording nits, feel free to ignore if they're not helpful. Looks good!

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 2 above)
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:311]: @BinaryFissionGames What do you think about this wording? It's a nit, but I think this is a bit clearer.

**Review by evan-bradley — COMMENTED** [review]
Thanks for submitting a fix for this, and sorry for the long delay. Could you explain the flow of what is returned to me? If we execute `return errors.New("...")`, is `err` set to the return value in the deferred functions?

File: cmd/opampsupervisor/supervisor/supervisor.go (hunk 2 above)
Comment (BinaryFissionGames) [reply cmd/opampsupervisor/supervisor/supervisor.go]: I must have dropped it when rebasing at some point, I definitely meant to take these suggestions.

**Comment by BinaryFissionGames** [comment]
> Thanks for submitting a fix for this, and sorry for the long delay. Could you explain the flow of what is returned to me? If we execute `return errors.New("...")`, is `err` set to the return value in the deferred functions?

Yes, that's exactly right. In the defer statements, `err` will be the value that will be returned (the errors.New value). Then, the defers can further modify that value in cases where it has it's own error.

Here's an example: https://go.dev/play/p/hT6HZSK62NH

**Review by evan-bradley — APPROVED** [review]

**Comment by evan-bradley** [comment]
>Yes, that's exactly right. In the defer statements, err will be the value that will be returned (the errors.New value). Then, the defers can further modify that value in cases where it has its own error.

Great, thank you. That's straightforward, just hadn't seen that pattern before.

**Review by djaglowski — APPROVED** [review]

**Pull request merged**